Mid-term Evaluation: Backend

bphenriques opened the pull request:
- Medicine Manager with registries manager, statistics computation and notifications handler. It is ready for GraphView.
- PlanMyTrip: Add/remove trip and items. Also supports query.
- Rest API Integration. Is extensible for more endpoints.
- Centralised user settings.
- Unit tests for the above items.
- Started Front-end however it is not ready. It will be developed later this week.

kangaroo5383 commented:
This looks good overall.
:+1: for using cocoa pods for dependency management.

Some pattern notes:
1. It looks like you're reusing a single background context for your manage objects, this is not ideal as it makes your architecture rigid. Usually the view controller manage what context the information that's modified in it is saved to.
For more info: https://developer.apple.com/library/ios/documentation/DataManagement/Conceptual/CoreDataSnippets/Articles/stack.html#//apple_ref/doc/uid/TP40008283-SW2
You can check out something like magical records https://github.com/magicalpanda/MagicalRecord for some convenience around managing that if it comes a time where the app should need a more flexible implementation.

kangaroo5383 commented on malaria-ios/Managers/MedicineHelpers/MedicineManager.swift:
```diff
@@ -0,0 +1,109 @@
+        newMed.notificationManager.scheduleNotification(fireDate)
+
+        UserSettingsManager.setBool(UserSetting.DidConfiguredMedicine, true)
+        UserSettingsManager.syncronize()
```
you don't actually need synchronize nowadays, writing to disk operation is better left to system optimization.

bphenriques replied:
"The synchronize method, which is automatically invoked at periodic intervals, keeps the in-memory cache in sync with a user’s defaults database.".

At this case, is it safe? Should I call syncronize only when the application exists?

kangaroo5383 replied:
Yeah during the course of the application cycle, it's saved frequently when modified.  Only calling on exit is a good strategy, some would argue that's not even necessary because how quickly that's saved - this is a minor point however.

kangaroo5383 commented on malaria-ios/Managers/UserSettingsManager.swift:
```diff
@@ -0,0 +1,29 @@
+    /// Sets an UserSetting
+    /// :param: `UserSetting`: the setting to be changed
+    /// :param: `Bool`: value
+    class func setBool(key: UserSetting, _ value: Bool){
```
Instead of these helper methods, you should try just implementing
`UserSettingsManager.didConfiguredMedicine` / `UserSettingsManager.setDidConfiguredMedicine` etc that way you can consolidate your string keys here as well.  Or extend `NSUserDefaults` so you can just make calls like `NSUserDefaults.standardUserDefaults().setDidConfigureMedicine` etc.  Same idea.  Let me know if that make sense.

bphenriques replied:
Yes, it makes! However I would need to create a `get`and a `set`for each user setting. However, since there is only that setting for now, I will follow your first suggestion.

kangaroo5383 replied:
Yeah, it might seem a bit redundant but the benefit is that you can easily see all the different user defaults that your system is using (with autocomplete :+1: ), and also all the keys you're using so there won't be any overlaps.  It solves the same concern as the `UserSetting` you had in consolidating the strings in one place, but with one less layer of abstraction.

kangaroo5383 commented on malaria-ios/ViewControllers/DailyStatsTableViewController.swift:
```diff
@@ -0,0 +1,37 @@
+            let cell = tableView.dequeueReusableCellWithIdentifier("medicineLastTakenCell") as! UITableViewCell
+            return cell
+        }else if row == 1 {
+            let cell = tableView.dequeueReusableCellWithIdentifier("dosesCell") as! UITableViewCell
```
Instead of using magic numbers here to check you can use enums to give them meaning such as the following for clarity.  The last count one would also give you the proper value for numberOfRowsInSection so it's a simpler change if you have to add more rows or change their order.

`DailyStatsMedicineLastTakenCell`
`DailyStatsDosesCell`
`DailyStatsCellCount`

bphenriques replied:
That's a good idea, thank you. 

The front-end was interrupted therefore the code isn't as organized as I would like it to be. I will resume development after fixing the current pull-request.

bphenriques replied:
https://github.com/bphenriques/malaria-app-ios/blob/bf3c028222651d4a453ed0d4b0589b56056c0761/malaria-ios/ViewControllers/DailyStatsTableViewController.swift

kangaroo5383 replied:
Yes, this is a good implementation as well if you don't actually need the sections - removing the extra complexity.

bphenriques commented:
Hello,

Thank you for the feedback! Regarding Core Data, the concept is still unfamiliar to me and after reading your comment, I researched more but brought more questions.

1. Since we are not using large amounts of data at a time what will be the advantage of using multiple contexts? Will the application take profit of it since the data model is simple and mostly single-threaded?

2. We can have at most two working threads that I know of: The main thread and another while syncing REST API, therefore should the app have two contexts? Main context and background context? How many contexts should I have and how should one manage them and why they should be handled that way? 

I took a look in MagicalRecords and in really simplifies the code and extends the current functionality for the future, and it's thread safe. Do you suggest moving to it?

kangaroo5383 commented:
I find Magical Records to be a good tool in simplifying some of the common operations, if it's not too difficult to integrate it, it would be good for maintenance and extensibility.

In the case 2 above, you can't have just two context - one main thread and one in the background - because at some point they will have different information and diverge.  The most straightforward way to manage that is to have the policy to always be creating new context from the background context and do editing with that new context.  when save happens, the changes is then merged back.  That would be the general principle, let me know if you have issues / questions when you get to the implementation.

bphenriques commented:
Thank you. I'll look into it as soon as possible. I am just getting things ready for a demo this weekend to have early feedback.

bphenriques commented:
Hello Jessica,

At the moment, in the backend I am saving the data whenever I need them to be persistent. However, by looking on magical records and with what you said regarding "policy to always be creating new context from the background context and do editing with that new context.", I am not sure if my current architecture fits. i.e. saving everytime I change an object and not creating some sort of transaction (e.g. only save when I dismiss a view). If it doesn't fit, I am not quite sure how I should approach the problem.

Maybe I am not making myself clear. If so, let me know. I would like to talk with you and share some of my concerns and maybe you could give me some more details regarding the overall handling of contexts while looking at concrete simple examples of the code (e.g. TripsManager).

(It's not that the framework it self is hard, is that maybe the app should have been already been developed with Magical Record in mind. But well, that's why I am here learning :) )

My contact: brunoaphenriques@gmail.com

kangaroo5383 commented:
@bphenriques Sure I think talking over it would be most effective.  I will email you with tomorrow to set up a time and provide you with some sample code.  This is not blocking your progress on other parts of the app correct?

bphenriques commented:
@kangaroo5383 Ok! On one hand yes, I could be developing other parts of the app but, on the other hand, this was already planned so don't worry, everything is on schedule and since this will improve the overall architecture and "close" the backend, I will make this a priority.

bphenriques commented:
Hello,

Regarding CoreData:
I haven't been able to integrate with MagicalRecord (or any other framework...) due to some bugs that can be observed in the experimental branch (pod install -> Run Tests). However, integration is simple because the code is adapted to receive an NSManagedObjectContext therefore it is something to be reviewed later. As a solution and since it wasn't that hard to extend the code (just had to understand what I was doing), I implemented manually the handling of multiple contexts. Every viewController has got his own NSManagedObjectContext of type privateQueue that has as parent context the mainQueue (that is never directly accessed). When a background context is saved, the main context merges with it and then saves the changes to persistent storage.

One more thing: Fixed lightweight migration  (would crash the user application if I change even the tiniest thing in the model) and removed an hack to test NSManagedObjectContext (the hack and the bug were both related). I think this was important to implement before submitting the app.

Regarding UI:
Prepared the app to be submitted to the AppStore. Added icon, launch screen, added alertViews, fixed consistency between fonts and colours; Fix TabBar items icons. I tested in iphone 4s, iphone 5, iphone 5s, iphone 6 and iphone 6s and looks good to me. Any feedback would be great! 

What is missing:
- GraphView;
- Test on a real device (I need to collaborate with Rosario to fix it): notifications and usability when using the app (for example increase the hit area of the buttons);
- When Swift 2 is released it would be interesting to use @testable: http://natashatherobot.com/swift-2-xcode-7-unit-testing-access/ instead of adding "public" to every (needed) class/method/var.

kangaroo5383 commented:
Re: migration - that's the expected behavior in a sense. When an app is
shipped with a model definition and then in the next version that model
definition is changed - you just have to create a migration. You only have
to create a migration with each shipped version. During development you
should do a content reset and work from the latest model definitions. I
hope that helps!
On Sun, Jul 12, 2015 at 5:53 PM Bruno Henriques <notifications@github.com>
wrote:

> Hello,
>
> Regarding CoreData:
> I haven't been able to integrate with MagicalRecord (or any other
> framework...) due to some bugs that can be observed in the experimental
> branch (pod install -> Run Tests). However, integration is simple because
> the code is adapted to receive an NSManagedObjectContext therefore it is
> something to be reviewed later. As a solution and since it wasn't that hard
> to extend the code (just had to understand what I was doing), I implemented
> manually the handling of multiple contexts. Every viewController has got
> his own NSManagedObjectContext of type privateQueue that has as parent
> context the mainQueue (that is never directly accessed). When a background
> context is saved, the main context merges with it and then saves the
> changes to persistent storage.
>
> One more thing: Fixed lightweight migration (would crash the user
> application if I change even the tiniest thing in the model) and removed an
> hack to test NSManagedObjectContext (the hack and the bug were both
> related). I think this was important to implement before submitting the app.
>
> Regarding UI:
> Prepared the app to be submitted to the AppStore. Added icon, launch
> screen, added alertViews, fixed consistency between fonts and colours; Fix
> TabBar items icons. I tested in iphone 4s, iphone 5, iphone 5s, iphone 6
> and iphone 6s and looks good to me. Any feedback would be great!
>
> What is missing:
>
>    - GraphView;
>    - Test on a real device (I need to collaborate with Rosario to fix
>    it): notifications and usability when using the app (for example increase
>    the hit area of the buttons);
>    - When Swift 2 is released it would be interesting to use @testable
>    <https://github.com/testable>:
>    http://natashatherobot.com/swift-2-xcode-7-unit-testing-access/
>    instead of adding "public" to every (needed) class/method/var.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/systers/malaria-app-ios/pull/7#issuecomment-120777445>
> .
>

bphenriques commented:
After todays dry-run with Rose and Nick for tomorrow's meeting with peace corps. Found a bug when switching the current pill that crashes the app. Among the issues, one of them was context synchronisation, as a solution, a new context is created in viewWillAppear method. Moreover, added a confirmation dialog when the user tries to change the current configuration.

rosariorobinson commented:
Has this been fixed?

bphenriques commented:
Yes. This pull request is old and there are lots of changes since then. This is the most recent one still pending.

https://github.com/systers/malaria-app-ios/pull/8
